### DISCUSSION TITLE
Add Custom directive to Relay Pagination types generated by codegen

### DISCUSSION THREAD

**Discussion by yamalameyooooo** [Netflix/dgs-framework#1589] [Q&A]
Hi, 
Currently i am using codegen to generate the Types related to relay pagination whereever i have added @connection directive.

Sample schema file:
```

directive @shareable repeatable on OBJECT | FIELD_DEFINITION

type Query {
      hello: MessageConnection
}

type Message @connection @shareable{
    name: String
}
```

my codegen configuration:
```
generateJava{
 ...
  typeMapping = ["MessageConnection": "graphql.relay.SimpleListConnection<Message>"]
}
```

the outcome of the schema in runtime(verified by the sdl too):

```
directive @shareable repeatable on OBJECT | FIELD_DEFINITION

type Query {
      hello: MessageConnection
}

type Message {
    name: String
}

"MessageConnection"
type MessageConnection {
  "Field edges"
  edges: [MessageEdge]
  "Field pageInfo"
  pageInfo: PageInfo
}

"MessageEdge"
type MessageEdge {
    "Field node"
    node: Message
    "Field cursor"
    cursor: String
}

"PageInfo"
type PageInfo {
    "Field hasPreviousPage"
    hasPreviousPage: Boolean!
    "Field hasNextPage"
    hasNextPage: Boolean!
    "Field startCursor"
    startCursor: String
    "Field endCursor"
    endCursor: String
}
```

As you can see, codegen created 3 types in runtime that is **MessageConnection, MessageEdge and PageInfo**

but i need my custom directive @shareable on these 3 directives too. like this:

```
"MessageConnection"
type MessageConnection @shareable {
  "Field edges"
  edges: [MessageEdge]
  "Field pageInfo"
  pageInfo: PageInfo
}

"MessageEdge"
type MessageEdge @shareable {
    "Field node"
    node: Message
    "Field cursor"
    cursor: String
}

"PageInfo"
type PageInfo @shareable {
    "Field hasPreviousPage"
    hasPreviousPage: Boolean!
    "Field hasNextPage"
    hasNextPage: Boolean!
    "Field startCursor"
    startCursor: String
    "Field endCursor"
    endCursor: String
}
```

How can i and what to configure in codegen configuration to get something like this above in this configuration?

@srinivasankavitha can you guide me in this problem statement, as my this is going to be in federated in a supergraph and there are other identical types too which is giving a clash. 
kindly let me know how can i add this directive in the codegen generated relay pagination types

**Comment by srinivasankavitha** [comment]
Hi @yamalameyooooo - The use of `@connection` directive is a convenience added for very basic generation of relay types. If you need customization, you will need to avoid using `@connection` and manually define the types in the schema adding your `@shareable` directive. The framework currently does not provide any customizations when using `@connection` at the moment.

**Reply by yamalameyooooo** [reply]
@srinivasankavitha i added the types manually, and generated the corresponding classes too like Edge, PageInfo etc etc, 
i am at blank, how do i implement the logic now , is there any sample repo you can guide me to where i can see how manually everything is implemented? 
would save me a lot

**Comment by deweyjose** [comment]
Hi @srinivasankavitha,

@yamalameyooooo opened an [issue](https://github.com/deweyjose/graphqlcodegen/issues/114) in the maven graphqlcodegen plugin for DGS related to this topic. 

I think the `@connection` directive is super useful and saves a ton of boilerplate schema. However the dynamic schema for connection related types won't work in Federation today as they need to be `@sharable` - especially the `PageInfo` type. 

If folks have advanced Apollo schema composition pipelines they could inject additional directives into the schema JIT so that it's there for Apollo and the Federated ecosystem. Not everyone has that infrastructure built though. 

I think we could drive customization of the dynamic connection types in an elegant way either via additional schema directives or via Spring configuration. 

Spitballing here but... 

perhaps something like this where we're explicit about what to include with dynamic connection types:

`type foo @connection(inherit = ["shareable", "xyz"]) @shareable @xyz(bar="foo") { }`

or maybe the rule is to just propagate them all:

`type foo @connection @shareable @xyz(bar="foo") { }`

With Spring config maybe?

```
dgs:
  connection:
    custom-directives:
      - sharable
      - xyz:
          foo: bar
```

**Reply by srinivasankavitha** [reply]
Makes sense, I like the approach of having additional spring configuration might be the better approach. I would hesitate to add more custom directives that could end up in the schema. We try not to add coupling between  generalgraphql concepts and federation so that they can be used independently. Your idea of adding the `@shareable` based on config would fit nicely with that I think.

**Reply by yamalameyooooo** [reply]
@srinivasankavitha this would be a great to this feature as @deweyjose mentioned mentioned below:
```
dgs:
  connection:
    custom-directives:
      - sharable
      - tag:
          name: "public"
```
and to have in spring configuration, this will solve a lot of boiler plate code and the developer would be more concentrated on the service logic rather than implementing things which are already present or have been done, plus this makes working with federation super easy
